### PR TITLE
Implement image pull secret propagation

### PR DESCRIPTION
**Pull request by mmorhun**
This PR adds functionality to Image Controller that creates a `RemoteSecret` with application and component labels in order to make image pull secret available in all environments of the application.
Note, the image pull secrets gets linked to `default` service account in all environments automatically.

Also, `pull-secret` field was in image annotation was deleted.

<details>
<summary>Example of RemoteSecret after creation and credentials upload</summary>

```yaml
apiVersion: appstudio.redhat.com/v1beta1
kind: RemoteSecret
metadata:
  creationTimestamp: "2023-07-10T08:34:47Z"
  finalizers:
  - appstudio.redhat.com/linked-objects
  - appstudio.redhat.com/secret-storage
  generation: 1
  labels:
    appstudio.openshift.io/application: go-application-sample
    appstudio.openshift.io/component: go-component-sample
    secretuploaded: "true"
  name: go-component-sample-pull
  namespace: test2
  ownerReferences:
  - apiVersion: appstudio.redhat.com/v1alpha1
    kind: Component
    name: go-component-sample
    uid: e4575294-19cf-4190-90b2-5b8df3ea242a
  resourceVersion: "25653356"
  uid: 076a0921-ef56-4fc9-9cbd-a0247d9b9c64
spec:
  secret:
    linkedTo:
    - serviceAccount:
        as: secret
        managed:
          generateName: ""
          name: ""
        reference:
          name: appstudio-pipeline
    name: go-component-sample-pull
    type: kubernetes.io/dockerconfigjson
status:
  conditions:
  - lastTransitionTime: "2023-07-10T08:34:47Z"
    message: ""
    reason: DataFound
    status: "True"
    type: DataObtained
  - lastTransitionTime: "2023-07-10T08:34:47Z"
    message: ""
    reason: Injected
    status: "True"
    type: Deployed
```

</details>

### CODE REVIEW

File: Dockerfile [hunk 1]
```diff
@@ -1,5 +1,5 @@
 # Build the manager binary
-FROM golang:1.18 as builder
+FROM golang:1.20 as builder
```
Comment (Michkov) [inline Dockerfile:2]: nit: we should rather use images provided by Redhat (not related but when you already touching this)

File: controllers/component_image_controller.go [hunk 2]
```diff
@@ -331,32 +347,114 @@ func (r *ComponentReconciler) reportError(ctx context.Context, component *appstu
...
+			},
+			Spec: remotesecretv1beta1.RemoteSecretSpec{
+				Secret: remotesecretv1beta1.LinkableSecretSpec{
+					Name: remoteSecretName,
+					Type: corev1.SecretTypeDockerConfigJson,
+					LinkedTo: []remotesecretv1beta1.SecretLink{
+						{
+							ServiceAccount: remotesecretv1beta1.ServiceAccountLink{
+								Reference: corev1.LocalObjectReference{
+									Name: BuildPipelineServiceAccountName,
```
Comment (Michkov) [inline controllers/component_image_controller.go:416]: This pull secret should not be linked to appstudio-pipeline service account, since there is already linked push secret which is using same repo path.

File: controllers/component_image_controller.go [hunk 3]
```diff
@@ -236,24 +243,33 @@ func (r *ComponentReconciler) Reconcile(ctx context.Context, req ctrl.Request) (
...
-				} else {
-					log.Info(fmt.Sprintf("Updated image registry pull secret %s for Component", pullRobotAccount.Name), l.Action, l.ActionUpdate)
 				}
+				log.Info(fmt.Sprintf("Prepared image registry pull secret %s for Component", pullRobotAccount.Name), l.Action, l.ActionUpdate)
+
+				// Propagate the pull secret into all environments
+				if err := r.EnsureComponentPullSecretRemoteSecret(ctx, component, pullSecretName); err != nil {
+					return ctrl.Result{}, err
+				}
+				if err := r.CreateRemoteSecretUploadSecret(ctx, component.Namespace, pullSecretName, pullSecretData); err != nil {
```
Comment (Michkov) [inline controllers/component_image_controller.go:269]: do we need to keep the '-pull' secret in the namespace when it was uploaded already?

File: controllers/component_image_controller.go (hunk 2 above)
Comment (Michkov) [reply controllers/component_image_controller.go:416]: Currently user workload is running under `default` SA.

**Comment by Michkov** [comment]
@skabashnyuk Please check the output remoteSecret from description, should it look like this? I can see errors in my testing env in remotesecrets controller:
```
{"level":"error","ts":"2023-07-10T10:10:04.215Z","caller":"controller/controller.go:329","msg":"Reconciler error","controller":"remotesecret","controllerGroup":"appstudio.redhat.com","controllerKind":"RemoteSecret","RemoteSecret":{"name":"devfile-sample-code-with-quarkus-pull","namespace":"test"},"namespace":"test","name":"devfile-sample-code-with-quarkus-pull","reconcileID":"c2e97ff5-0446-4522-ad2a-0d36de69eefd","error":"failed to persist the stage result condition in the status after the stage secret-deployment: Operation cannot be fulfilled on remotesecrets.appstudio.redhat.com \"devfile-sample-code-with-quarkus-pull\": the object has been modified; please apply your changes to the latest version and try again","stacktrace":"sigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).reconcileHandler\n\t/go/pkg/mod/sigs.k8s.io/controller-runtime@v0.14.6/pkg/internal/controller/controller.go:329\nsigs.k8s.io/controller-runtime/pkg/internal/controller.(*Controller).processNextWorkItem\n\t/go/pkg/...
```

File: Dockerfile (hunk 1 above)
Comment (mmorhun) [reply Dockerfile:2]: Let's do it in a separate PR

File: controllers/component_image_controller.go (hunk 2 above)
Comment (mmorhun) [reply controllers/component_image_controller.go:416]: Thanks, fixed.

File: controllers/component_image_controller.go [hunk 4]
```diff
@@ -331,32 +347,114 @@ func (r *ComponentReconciler) reportError(ctx context.Context, component *appstu
...
+				"appstudio.redhat.com/upload-secret": "remotesecret",
+			},
+			Annotations: map[string]string{
+				"appstudio.redhat.com/remotesecret-name": remoteSecretName,
+			},
+		},
+		StringData: secretData,
+	}
+	if err := r.Client.Create(ctx, uploadSecret); err != nil {
+		log.Error(err, fmt.Sprintf("failed to create remote secret: %v", uploadSecret.Name), l.Action, l.ActionAdd, l.Audit, "true")
```
Comment (tkdchen) [inline controllers/component_image_controller.go:451]: How about `s/remote secret/short live upload secret/`?

File: Dockerfile (hunk 1 above)
[the remark this answers is not in the document]
Comment (mmorhun) [reply Dockerfile:2]: Because remote-secret API required go `1.19`

**Review by tkdchen — COMMENTED** [review]
:+1: 

Additional thought:

nitpick: label value "remotesecret" could be another constant provided by remote-secret.

/cc @skabashnyuk What do you think remote-secret provides API to create RemoteSecret object and a short live upload secret?

**Review by tkdchen — APPROVED** [review]

File: controllers/component_image_controller.go [hunk 5]
```diff
@@ -37,13 +37,19 @@ import (
...
 )
 
 const (
 	ImageAnnotationName         = "image.redhat.com/image"
 	GenerateImageAnnotationName = "image.redhat.com/generate"
 
 	ImageRepositoryFinalizer = "image-controller.appstudio.openshift.io/image-repository"
+
+	ApplicationNameLabelName = "appstudio.openshift.io/application"
+	ComponentNameLabelName   = "appstudio.openshift.io/component"
```
[the remark this answers is not in the document]
Comment (mmorhun) [reply controllers/component_image_controller.go]: Updated

**Comment by tkdchen** [comment]
> > /cc @skabashnyuk What do you think remote-secret provides API to create RemoteSecret object and a short live upload secret?
> 
> @tkdchen I'm not fully sure that I understand the question correctly. Could you rephrase it?

I was thinking if it is worth for remote-secret providing API to encapsulate the details of creating RemoteSecret and the short live upload Secret.

**Review by Michkov — APPROVED** [review]
